File: server/usecases/DashboardUseCase.ts
import { IProxyRepository, IDomainRepository, IBugRepository } from "../repositories/interfaces";
import { logger } from "../utils/logger";

export interface DashboardStats {
  proxies: number;
  domains: number;
  bugs: number;
}

export class DashboardUseCase {
  constructor(
    private proxyRepo: IProxyRepository,
    private domainRepo: IDomainRepository,
    private bugRepo: IBugRepository
  ) {}

  async getStats(): Promise<DashboardStats> {
    logger.debug("Retrieving dashboard stats", "DashboardUseCase");
    const totalProxies = await this.proxyRepo.count();
    const totalDomains = await this.domainRepo.count();
    const totalBugs = await this.bugRepo.count();

    return {
      proxies: totalProxies,
      domains: totalDomains,
      bugs: totalBugs,
    };
  }
}

File: checker2-go/internal/module/dashboard/usecase/dashboard_usecase.go
package usecase

import (
	"context"

	"github.com/hidessh99/sub-cf-vpn/checker2-go/internal/module/dashboard/port"
	sharedPort "github.com/hidessh99/sub-cf-vpn/checker2-go/internal/shared/port"
)

type DashboardUseCase interface {
	GetStats(ctx context.Context) (*port.DashboardStats, error)
}

type dashboardUseCase struct {
	statsProvider port.StatsProvider
	log           sharedPort.Logger
}

func NewDashboardUseCase(statsProvider port.StatsProvider, log sharedPort.Logger) DashboardUseCase {
	return &dashboardUseCase{
		statsProvider: statsProvider,
		log:           log,
	}
}

func (u *dashboardUseCase) GetStats(ctx context.Context) (*port.DashboardStats, error) {
	u.log.Debug("Retrieving dashboard stats", "DashboardUseCase")

	proxiesCount, err := u.statsProvider.ProxyCount(ctx)
	if err != nil {
		u.log.Error("Database error during proxies count retrieval", err, "DashboardUseCase")
		return nil, err
	}

	domainsCount, err := u.statsProvider.DomainCount(ctx)
	if err != nil {
		u.log.Error("Database error during domains count retrieval", err, "DashboardUseCase")
		return nil, err
	}

	bugsCount, err := u.statsProvider.BugCount(ctx)
	if err != nil {
		u.log.Error("Database error during bugs count retrieval", err, "DashboardUseCase")
		return nil, err
	}

	return &port.DashboardStats{
		Proxies: proxiesCount,
		Domains: domainsCount,
		Bugs:    bugsCount,
	}, nil
}
